src/librustc_mir/transform/type_check.rs

```diff
@@ -1068,81 +1071,132 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
+                                continue;
+                            }
+                        };
+                        let op_ty = match op {
+                            Operand::Consume(lv) => {
+                                let lv_ty = lv.ty(mir, tcx).to_ty(tcx);
+                                if self.is_adt_constructor {
+                                    self.normalize(&lv_ty, location)
+                                } else {
+                                    lv_ty
                                 }
+                            }
+                            Operand::Constant(c) => c.ty,
+                        };
+                        if let Err(terr) =
+                            self.sub_types(op_ty, field_ty, location.at_successor_within_block())
+                        {
+                            span_mirbug!(
+                                self,
+                                rv,
+                                "{:?} is not a subtype of {:?}: {:?}",
+                                op_ty,
+                                field_ty,
+                                terr
+                            );
                         }
-                    }
+                    },
                 }
             }
+
+            Rvalue::Repeat(op, _) => {
+                // FIXME: extract this to a method
+                let op_ty = match op {
+                    Operand::Consume(lv) => {
+                        let lv_ty = lv.ty(mir, tcx).to_ty(tcx);
+                        if self.is_adt_constructor {
+                            self.normalize(&lv_ty, location)
+                        } else {
+                            lv_ty
+                        }
+                    }
+                    Operand::Constant(c) => c.ty,
+                };
+
+                let trait_ref = ty::TraitRef {
+                    def_id: tcx.lang_items().copy_trait().unwrap(),
+                    substs: tcx.mk_substs_trait(op_ty, &[]),
+                };
+
+                self.prove(trait_ref, location);
+            }
+
+            Rvalue::NullaryOp(_, ty) => {
+                let trait_ref = ty::TraitRef {
+                    def_id: tcx.lang_items().sized_trait().unwrap(),
+                    substs: tcx.mk_substs_trait(ty, &[]),
+                };
+
+                self.prove(trait_ref, location);
+            }
+
             // FIXME: These other cases have to be implemented in future PRs
             Rvalue::Use(..) |
-            Rvalue::Repeat(..) |
             Rvalue::Ref(..) |
             Rvalue::Len(..) |
             Rvalue::Cast(..) |
             Rvalue::BinaryOp(..) |
             Rvalue::CheckedBinaryOp(..) |
             Rvalue::UnaryOp(..) |
-            Rvalue::Discriminant(..) |
-            Rvalue::NullaryOp(..) => {}
+            Rvalue::Discriminant(..) => {}
         }
     }
 
+    fn prove(&mut self, trait_ref: ty::TraitRef<'tcx>, location: Location) {
```

Comment:
So I realize now we should structure this a bit differently. Let's call this `prove_trait_implemented`. But we will implement it in terms of a new function, `prove_predicates`, which proves that a given set of predicates are true.

```rust
fn prove_predicate(&mut self, predicates: &[ty::Predicate<'tcx>], location: Location) {
    self.fully_perform_op(location.at_self(), |this| {
        let cause = this.misc(this.last_span);
        let obligations = predicates.iter().map(|&p| traits::Obligation::new(cause, this.param_env, p));
        Ok(InferOk { value: (), obligations: vec![obligation] })
    }).unwrap()
}
```

This is kind of a cute hack ---- it leverages the fact that `fully_perform_op` expects to be given back a list of predicates to prove. The actual "operation" is just to trivially succeed and pass back the predicates.

Then we can rewrite `prove_trait_ref` as follows:

```rust
fn prove_trait_ref(&mut self, trait_ref: ty::TraitRef<'tcx>, location: Location) {
    self.prove_predicates(&[ty::Predicate::TraitRef(trait_ref.to_poly_trait_predicate())], location);
}
```

Reply:
Done!